### PR TITLE
Audit DB logic in ODIS 2.0.0 & standardize signer metering

### LINKED ISSUE #9795: Audit db wrapper code + error handling
The Signer DB code needs to be thoroughly reviewed before we launch ODIS 2.0.0. In particular, we've seen that some tests may be flakey due to a potential hanging DB operation. The output of this ticket is a PR with any fixes/improvements to this part of the codebase, with the goal of identifying and fixing the source of this flakiness. 

Additionally, this PR should employ the [meter](https://github.com/celo-org/celo-monorepo/blob/79c6bdbe530cf872de6a63d0567a407172298cc6/packages/phone-number-privacy/signer/src/common/web3/contracts.ts#L23) helper function in any DB operations where it is not already being used, and should probably be taken by the same person who takes #9796

### LINKED ISSUE #9796: ensure meter(...) helper function is used everywhere in Signer
The [meter](https://github.com/celo-org/celo-monorepo/blob/79c6bdbe530cf872de6a63d0567a407172298cc6/packages/phone-number-privacy/signer/src/common/web3/contracts.ts#L23) helper function is currently being used in many places but not everywhere in the signer. For consistency, we should use this function instead of the more verbose metering code that is currently used in places such as https://github.com/celo-org/celo-monorepo/blob/79c6bdbe530cf872de6a63d0567a407172298cc6/packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts#L61

It will probably make sense to batch this work with #9795 as much of the old metering code is used in db wrapper functions

### PR DESCRIPTION
**Pull request by eelanagaraj**
### Changes
- uses `meter` function throughout the signer
  - moves the `meter` function to `metrics` (which felt like a better home than web3/contracts)
- added a couple of signer DB helper functions to make the queries slightly more readable and reduce some duplicate query code
  - added these to `signer/src/common/database/utils.ts` -- happy to move this elsewhere, but the `database` file directly didn't seem like a great fit
  - `tableWithLockForTrx` -- very open to other naming suggestions here
 - some minor DB fixes/optimizations where possible (i.e. for disabling a new domain, directly creating this domain as disabled instead of inserting and then updating)
 - fixed a couple of response edge cases when rolling back updates in the (LEGACY_)PNP_SIGN endpoints (see comment on this), and added a few unit test cases for this
 - added a small fix for our signer timeout logic + a test; also adds a TODO & ticket to review this more thoroughly (that felt out of scope for this PR, but we should audit the timeout/API response logic more thoroughly for sure)
 - use `toStrictEqual` instead of `toMatchObject` in our tests

### Tested
- added a couple of unit test cases for weird edge cases where the DB should be rolled back (signer PNP & legacyPNP)

### Related issues

- Fixes celo-org/celo-monorepo#9795 
- Fixes celo-org/celo-monorepo#9796

### CODE REVIEW

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts [hunk 1]
```diff
@@ -82,37 +85,41 @@ export class PnpSignAction implements Action<SignMessageRequest> {
...
-          this.config.keystore.keys.phoneNumberPrivacy.latest,
-        name: DefaultKeyName.PHONE_NUMBER_PRIVACY,
-      }
-
       // Compute signature inside transaction so it will rollback on error.
-
+      // Since it's possible that the trx was already rolled back in
+      // checkAndUpdateQuotaStatus, catch and handle any additional errors thrown,
+      // otherwise the signer will hang.
       try {
```
Comment (eelanagaraj) [inline packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: Turns out that errors thrown after the rollback do not throw (+ is not caught) by the outer error handling function (since calling rollback explicitly already causes the promise to reject; by default an error that triggers an auto-rollback also throws an error, but calling trx.rollback() (without passing in an error) doesn't). This means we need to send responses in this function itself otherwise the signer doesn't respond anything. An alternative here is having checkAndUpdateQuotaStatus return a boolean indicating whether the outer function _should_ rollback and then rolling back at the end, but that feels just as messy/tricky tbh :-/ Open to other ideas here!

File: packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts [hunk 2]
```diff
@@ -102,33 +94,32 @@ export async function updateDomainStateRecord<D extends Domain>(
...
-    // Consider doing this as part of DB audit ticket (https://github.com/celo-org/celo-monorepo/issues/9795)
-    const result = await getDomainStateRecord(db, domain, logger, trx)
+  return meter(
+    async () => {
+      const hash = domainHash(domain).toString('hex')
+      logger.debug({ hash, domain, domainState }, 'Update domain state')
+      // Check whether the domain is already in the database.
+      // TODO(2.0.0, refactor): Usage of this in the signature flow results in redudant queries of the current
+      // state. It would be good to refactor this to avoid making more than one SELECT.
+      // Consider doing this as part of DB audit ticket (https://github.com/celo-org/celo-monorepo/issues/9795)
```
Comment (eelanagaraj) [inline packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts]: The only ways I could come up with to avoid this second select would be moving the nonce check -> `checkAndUpdateQuotaStatus` or else checking against the value of `createEmptyOrReturn`, both of which I think are a bit messy & introduce more confusion/complexity so opted to keep as is for now. I'd recommend we remove this TODO in favor of a longer-term ticket that we deprioritize for 2.0.0.

File: packages/phone-number-privacy/signer/src/common/database/utils.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    default:
+      throw new Error('Unknown database label provided')
+  }
+
+  Counters.databaseErrors.labels(label).inc()
+  logger.error({ err }, errorMsg)
+  throw err
+}
+
+export function tableWithLockForTrx(baseQuery: Knex.QueryBuilder, trx?: Knex.Transaction) {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/common/database/utils.ts:36]: Nice!

File: packages/phone-number-privacy/signer/src/common/controller.ts [hunk 4]
```diff
@@ -11,18 +11,24 @@ export class Controller<R extends OdisRequest> {
...
-        await this.action.perform(session)
-      }
-    } catch (err) {
-      response.locals.logger.error(
-        { error: err },
-        `Unknown error in handler for ${this.action.io.endpoint}`
-      )
-      this.action.io.sendFailure(ErrorMessage.UNKNOWN_ERROR, 500, response)
-    }
+    await meter(
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/common/controller.ts:14]: 🚀

File: packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts [hunk 5]
```diff
@@ -1118,6 +1118,87 @@ describe('legacyPNP', () => {
...
+          const res = await sendRequest(req, authorization, SignerEndpoint.LEGACY_PNP_SIGN)
+
+          expect(res.status).toBe(500)
+          expect(res.body).toStrictEqual<SignMessageResponseFailure>({
+            success: false,
+            version: res.body.version,
+            performedQueryCount: performedQueryCount,
+            totalQuota: expectedQuota,
+            blockNumber: testBlockNumber,
+            error: ErrorMessage.KEY_FETCH_ERROR,
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts]: leaving this q for myself: why is this the default?

File: packages/phone-number-privacy/signer/src/server.ts [hunk 6]
```diff
@@ -135,15 +135,15 @@ export function startSigner(
...
-  addMeteredSignerEndpoint(SignerEndpoint.PNP_SIGN, pnpSign.handle.bind(pnpSign))
-  addMeteredSignerEndpoint(SignerEndpoint.PNP_QUOTA, pnpQuota.handle.bind(pnpQuota))
-  addMeteredSignerEndpoint(SignerEndpoint.DOMAIN_QUOTA_STATUS, domainQuota.handle.bind(domainQuota))
-  addMeteredSignerEndpoint(SignerEndpoint.DOMAIN_SIGN, domainSign.handle.bind(domainSign))
-  addMeteredSignerEndpoint(SignerEndpoint.DISABLE_DOMAIN, domainDisable.handle.bind(domainDisable))
-
-  addMeteredSignerEndpoint(SignerEndpoint.LEGACY_PNP_SIGN, legacyPnpSign.handle.bind(legacyPnpSign))
-  addMeteredSignerEndpoint(
+  logger.info('Right before adding meteredSignerEndpoints')
+  addEndpointWithTimeout(SignerEndpoint.PNP_SIGN, pnpSign.handle.bind(pnpSign))
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/server.ts:163]: It looks like we're no longer measuring the endpoint latency..did this get moved elsewhere?

File: packages/phone-number-privacy/signer/src/server.ts [hunk 7]
```diff
@@ -56,33 +57,32 @@ export function startSigner(
...
           Counters.timeouts.inc()
-          childLogger.warn(`Timed out after ${config.timeout}ms`)
+          errorMsg = ErrorMessage.TIMEOUT_FROM_SIGNER
+          errToLog = new Error(errorMsg)
         }
-      })
-      .finally(end)
-  }
+        childLogger.error({ errToLog })
+        res.status(500).json({
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/server.ts]: Can we use sendFailure() here?

Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: Hmm I see we don't have access to that here. Maybe if we were to pass addEndpointWithTimeout the Controllers as params rather than the handler functions?

File: packages/phone-number-privacy/signer/test/integration/pnp.test.ts [hunk 8]
```diff
@@ -903,6 +944,87 @@ describe('pnp', () => {
...
+          const res = await sendRequest(req, authorization, SignerEndpoint.PNP_SIGN)
+
+          expect(res.status).toBe(500)
+          expect(res.body).toStrictEqual<SignMessageResponseFailure>({
+            success: false,
+            version: res.body.version,
+            performedQueryCount: performedQueryCount,
+            totalQuota: expectedQuota,
+            blockNumber: testBlockNumber,
+            error: ErrorMessage.KEY_FETCH_ERROR,
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/pnp.test.ts]: I must be missing something, but from what I can tell this should be ErrorMessage.UNKNOWN_ERROR bc the error would be caught only be the outer-most catch block. Could you help me see what's actually happening here?

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts [hunk 9]
```diff
@@ -68,8 +69,10 @@ export class PnpSignAction implements Action<SignMessageRequest> {
           // should be based intelligently off of the warning messages returned by signers
           // (https://github.com/celo-org/celo-monorepo/issues/9836)
         }
-        const { sufficient } = await this.quota.checkAndUpdateQuotaStatus(quotaStatus, session, trx)
-        if (!sufficient) {
+        // quotaStatus is updated in place
+        const quotaUpdateRes = await this.quota.checkAndUpdateQuotaStatus(quotaStatus, session, trx)
+        dbUpdated = quotaUpdateRes.dbUpdated
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: nit: it doesn't look like we're modifying this anywhere else. If I'm not missing something, could you just access this as quotaUpdateRes.dbUpdated or use a const?

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 1 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: What if we only call trx.rollback() when the error is a SIGNATURE_COMPUTATION_FAILURE (since that's the only time we need to catch it here) and let the auto-rollback take care of all other scenarios, which would send 500 responses when caught by the outer-most catch block?

Comment (alecps) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: I.e move `await trx.rollback()` into the if-statement on line 104 in the original code?

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts [hunk 10]
```diff
@@ -82,37 +85,41 @@ export class PnpSignAction implements Action<SignMessageRequest> {
...
-        name: DefaultKeyName.PHONE_NUMBER_PRIVACY,
-      }
-
       // Compute signature inside transaction so it will rollback on error.
-
+      // Since it's possible that the trx was already rolled back in
+      // checkAndUpdateQuotaStatus, catch and handle any additional errors thrown,
+      // otherwise the signer will hang.
       try {
+        const key: Key = {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: Could you clarify why we should move this into the try-catch?

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 1 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: Hmm okay I think I see what you mean now. My concern with this current approach is that we resolve the type of ErrorMessage we should return in a way that's difficult to reason about. My vote might be to rely on auto-rollback wherever possible and only use trx.rollback() when we are sending a response. There's definitely some nuance here so happy to jump on a call to discuss further

Comment (alecps) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: It does seem like using trx.rollback() should probably be avoided given this behavior though

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 10 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: nvm I think I understand this now

**Review by alecps — CHANGES_REQUESTED** [review]
Looks awesome!! Just left a few comments / questions mostly around the trx.rollback() stuff

File: packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts (hunk 2 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts]: Sounds good to me. Thanks for investigating!

File: packages/phone-number-privacy/signer/src/server.ts (hunk 7 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: this would also reduce the number of times we need to call .bind which is kinda nice

File: packages/phone-number-privacy/signer/src/server.ts (hunk 6 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/server.ts:163]: I moved this into the controller's `handle` function (per one of the TODOs -- but maybe I misunderstood); let me know if you think it's useful to keep this in the main function and not in the `handle` function and I can change it back though!

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 9 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: unfortunately it's not in scope for the later try/catch block (which is what checks if the db was updated when attempting to rollback or not); lmk if you think it's more readable to just `let` declare `quotaUpdateRes` before this block and assign it here, and then use `quotaUpdateRes.dbUpdated` and I can change it to that!

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 1 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: I think it could make sense to discuss this sync! Condensed from convo on slack but:

> My vote might be to rely on auto-rollback wherever possible and only use trx.rollback() when we are sending a response.

I think we're currently still trying to do this as much as possible; unless we remove the rollback earlier in the code (allowing fail open but not storing queries if performed queries are incremented), I can't see a way of not needing to manually handle possible later rollbacks during key fetching/signing :-/

File: packages/phone-number-privacy/signer/src/server.ts (hunk 7 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/server.ts]: From what I could find about Express error handling, it may be a good idea to try/catch & send a response in the outer-most endpoint code anyways in case something goes reaaally wrong (so we at least return the error). I would suggest we take another look at this when we 1) consider moving the timeout logic code -> controller/handler anyways (which should result in timeout errors being handled by the controller directly, leaving this outer try/catch for super unexpected other things), and 2) audit our endpoint error handling code (i.e. double check that we are returning responses in all cases, etc.). I created a ticket for (1) and a draft ticket for (2), but what do you think of that? Otherwise happy to do that as part of this ticket but I think we should do 1 & 2 in any case!!

File: packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts (hunk 2 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/common/database/wrappers/domainState.ts]: updated the TODO & linked a ticket for the future :)

File: packages/phone-number-privacy/signer/src/common/database/wrappers/request.ts [hunk 11]
```diff
@@ -51,15 +40,14 @@ export async function storeRequest(
...
-  } finally {
-    storeRequestMeter()
-  }
+  return meter(
+    async () => {
+      logger.debug({ request }, 'Storing salt request')
+      await requests(db).transacting(trx).insert(new Request(request)).timeout(DB_TIMEOUT)
+    },
+    [],
+    (err: any) => countAndThrowDBError(err, logger, ErrorMessage.DATABASE_INSERT_FAILURE),
```
Comment (eelanagaraj) [inline packages/phone-number-privacy/signer/src/common/database/wrappers/request.ts:49]: note: this is different than before but intentionally want this registered as an INSERT failure (not update) to match other behavior

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 1 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts:111]: Updated to now
- propagate DB update errors (when storing requests or incrementing query counts)
  - default to controller's error handling for DB & key signing errors (internally, we should be able to debug from logs, but the combiner will receive UNKNOWN_ERROR (500)'s)
- throw a SIGNATURE_COMPUTATION_ERROR when anything fails in that sig computation try/catch (previously we would check the error message, but I think this is a bit simpler while providing a generally helpful hint about where things went wrong)

File: packages/phone-number-privacy/signer/src/server.ts (hunk 7 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/server.ts]: resolving in favor of the audit ticket

File: packages/phone-number-privacy/signer/test/integration/pnp.test.ts [hunk 12]
```diff
@@ -1085,6 +1115,45 @@ describe('pnp', () => {
           ).toBe(performedQueryCount)
           expect(await getRequestExists(db, req, rootLogger(_config.serviceName))).toBe(false)
         })
+
+        it('Should return 500 on non-SIGNATURE_COMPUTATION_FAILURE error in sign', async () => {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/pnp.test.ts]: Optional nit: I found the way this is worded as non-SIGNATURE_COMPUTATION_FAILURE slightly confusing while reading this. While I see how it makes sense after re-reading the bls-cryptography-client code, I wonder if it might be clearer if we called this '...on generic error in sign' or something like that?  Whatever you think is best

File: packages/phone-number-privacy/signer/test/integration/pnp.test.ts [hunk 13]
```diff
@@ -1003,23 +1037,19 @@ describe('pnp', () => {
...
+          expect(res.status).toBe(500)
+          expect(res.body).toStrictEqual<SignMessageResponseFailure>({
+            success: false,
             version: res.body.version,
-            signature: expectedSignature,
-            performedQueryCount: performedQueryCount + 1,
-            totalQuota: expectedQuota,
-            blockNumber: testBlockNumber,
-            warnings: [ErrorMessage.FAILURE_TO_STORE_REQUEST],
+            error: ErrorMessage.UNKNOWN_ERROR,
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/pnp.test.ts:1048]: I wouldn't worry about this for this PR, but could we add a TODO to investigate returning a more specialized error here so we know (from the combiner's perspective) when a signer is having DB connectivity issues and can ping them if they don't realize themselves? This doesn't need to go into the refactor but would be nice to do eventually. Curious roughly how big of a lift you think it would be

File: packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts [hunk 14]
```diff
@@ -359,7 +359,7 @@ describe('legacyPNP', () => {
         const res = await sendRequest(req, authorization, SignerEndpoint.LEGACY_PNP_QUOTA)
 
         expect(res.status).toBe(200)
-        expect(res.body).toMatchObject<PnpQuotaResponseSuccess>({
+        expect(res.body).toStrictEqual<PnpQuotaResponseSuccess>({
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts:366]: I'm seeing the flake tracker flagged that the beforeEach hook for these tests sometimes times out. We could try bumping the timeout to jest.setTimeout(20000) and see if we keep seeing these warnings

File: packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts [hunk 15]
```diff
@@ -1313,6 +1302,45 @@ describe('legacyPNP', () => {
           ).toBe(performedQueryCount)
           expect(await getRequestExists(db, req, rootLogger(_config.serviceName))).toBe(false)
         })
+
+        it('Should return 500 on error non-SIGNATURE_COMPUTATION_FAILURE error in sign', async () => {
```
Comment (alecps) [inline packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts]: same nit comment as below on the naming of this test

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts [hunk 16]
```diff
@@ -100,30 +106,28 @@ export class PnpSignAction implements Action<SignMessageRequest> {
...
         const signature = await this.sign(
           session.request.body.blindedQueryPhoneNumber,
           key,
           session
         )
         this.io.sendSuccess(200, session.response, key, signature, quotaStatus, session.errors)
+        return
       } catch (err) {
+        session.logger.error({ err })
+        // Note that errors thrown after rollback will have no effect
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: Out of curiousity, what do you think would happen if we moved the rollback to line 130 after the response is sent but before we return? This might be slightly better on the off chance the response sending function throws (though it shouldn't)

File: packages/phone-number-privacy/signer/src/pnp/services/quota.ts [hunk 17]
```diff
@@ -37,9 +37,11 @@ export abstract class PnpQuotaService
         sufficient = true
       }
     } else {
-      if (await this.updateQuotaStatus(session, trx)) {
-        state.performedQueryCount++
-      }
+      await Promise.all([
+        storeRequest(this.db, session.request.body, session.logger, trx),
```
Comment (alecps) [inline packages/phone-number-privacy/signer/src/pnp/services/quota.ts:41]: So much cleaner 👌

**Review by alecps — APPROVED** [review]
Awesome work!! 😮‍💨

File: packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts (hunk 16 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts]: Ugh yeah I went back and forth a bit on this but not totally sure which way is better :-/ I don't love decrementing performedQueryCount before we know the rollback succeeded and figured that if sendFailure fails here, it would likely fail in the outer handler as well (maybe an unfair assumption), but maybe possible inconsistencies in received responses is better than not handling possible sendFailure errors 🤔 Happy to switch for now and we can keep this in mind

File: packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts (hunk 14 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/legacypnp.test.ts:366]: sure, will bump the timeout in this PR and add a TODO to check back on this as part of the timeout ticket (even if they're not actually related) if that sounds ok to you!

File: packages/phone-number-privacy/signer/test/integration/pnp.test.ts (hunk 13 above)
Comment (eelanagaraj) [reply packages/phone-number-privacy/signer/test/integration/pnp.test.ts:1048]: Can definitely add a TODO and agree that it isn't necessary pre-merge.

My thinking on this during the PR:
- the most "straight forward" seeming way (adding try/catch's to each block of code that we want to return specific errors) causes the code to become more complicated (nervous that we will very likely add in some weird bugs/edge cases in the future since we would again need to add in rollbacks for each of these errors and return)
- we could try to match on certain types of error messages in the outer handler (i.e. DB, key errors) and return specific responses; I started looking into this but then decided it would be a bit out of scope for this PR (it felt a bit more involved than it sounds as soon as I started actually implementing it)
- we could do something kinda weird like put all of the code that needs to rollback in the last try/catch and have a "most likely error" param that changes to reference the coming block and then handle the errors + rollback accordingly (though this is also pretty weird)

Created a ticket here: https://github.com/celo-org/celo-monorepo/issues/9882
